pkg/sypgp/sypgp.go

```diff
@@ -372,18 +373,37 @@ func GenKeyPair() (entity *openpgp.Entity, err error) {
 		return
 	}
 
-	fmt.Print("Generating Entity and OpenPGP Key Pair... ")
-	entity, err = openpgp.NewEntity(name, comment, email, conf)
+	pass, err := AskQuestionNoEcho("Enter encryption passphrase : ")
```

Comment:
Can we please move the asking to a function and handle the retries using that?

Something like this:

```
var errPassphraseMismatch = errors.New("passphrases do not match")
var errTooManyRetries = errors.New("too many retries while getting a passphrase")

func getPassphrase(retries int) (string, error) {
    ask := func() (string, error) {
        pass1, err := AskQuestinNoEcho("...")
        if err != nil {
            return "", err
        }

        pass2, err := AskQuestinNoEcho("...")
        if err != nil {
            return "", err
        }

        if pass1 != pass2 {
            return "", errPassphraseMismatch
        }

        return pass1, nil
    }

    for i := 0; i < retries; i++ {
        switch passphrase, err := ask(); err {
        case nil:
            // we got it!
           return passphrase, nil
        case errPassphraseMismatch:
            // retry
        default:
            // something else went wrong, bail out
            return "", err
        }
    }

    return "", errTooManyRetries
}
```

and then here you can do:

```
    passphrase, err := getPassphrase(3)
    if err != nil {
        return ...
    } 
```

Reply:
Resolved by [`
35271f8`](https://github.com/sylabs/singularity/pull/2886/commits/35271f8dfd516f0871a2b22f911a1a9347eeac3d).